<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Rodrigo Amadeu - rramadeu at ufl dot edu  Patricio Munoz  Horticultural Sciences Department - University of Florida" />

<meta name="date" content="2018-05-30" />

<title>AGHmatrix: An R package to compute relationship matrices for diploid and autopolyploid species</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">AGHmatrix: An R package to compute relationship matrices for diploid and autopolyploid species</h1>
<h4 class="author"><em>Rodrigo Amadeu - rramadeu at ufl dot edu </br> Patricio Munoz </br> Horticultural Sciences Department - University of Florida</em></h4>
<h4 class="date"><em>2018-05-30</em></h4>



<div id="overview" class="section level2">
<h2>Overview</h2>
<p>AGHmatrix software is an R-package to build relationship matrices using pedigree (A matrix) and/or molecular markers (G matrix) with the possibility to build a combined matrix of Pedigree corrected by Molecular (H matrix). The package works with diploid and autopolyploid Data.</p>
<p>If you are not familiar with <code>R</code>, we recommend the reading of vignette <a href="http://htmlpreview.github.io/?https://github.com/augusto-garcia/onemap/blob/master/inst/doc/Introduction_R.html">Introduction to R</a>.</p>
<center>
<p><em>Pedigree-based relationship matrix (A matrix)</em></p>
<table>
<thead>
<tr class="header">
<th></th>
<th>       </th>
<th align="center">Additive</th>
<th>       </th>
<th align="center">Non-Additive</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Diploid</strong></td>
<td>       </td>
<td align="center">Henderson (1976)</td>
<td>       </td>
<td align="center">Cockerham (1954)</td>
</tr>
<tr class="even">
<td><strong>Polyploid</strong></td>
<td>       </td>
<td align="center">Kerr (2012), Slater (2013)</td>
<td>       </td>
<td align="center"></td>
</tr>
</tbody>
</table>
</center>
<center>
<p><em>Molecular-based relationship matrix (G matrix)</em></p>
<table style="width:100%;">
<colgroup>
<col width="12%"></col>
<col width="16%"></col>
<col width="33%"></col>
<col width="5%"></col>
<col width="32%"></col>
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>       </th>
<th align="center">Additive</th>
<th>       </th>
<th align="center">Non-Additive</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Diploid</strong></td>
<td>       </td>
<td align="center">Yang (2010), VanRaden (2012)</td>
<td>       </td>
<td align="center">Su (2012), Vitezica (2013)</td>
</tr>
<tr class="even">
<td><strong>Polyploid</strong></td>
<td>       </td>
<td align="center">Slater (2016), VanRaden (2012)</td>
<td>       </td>
<td align="center">Slater (2016), Endelman (2018)</td>
</tr>
</tbody>
</table>
</center>
</div>
<div id="citation" class="section level2">
<h2>Citation</h2>
<p>To cite this R package:</p>
<p>Amadeu, R. R., C. Cellon, J. W. Olmstead, A. A. F. Garcia, M. F. R. Resende, and P. R. Muñoz. 2016. AGHmatrix: R Package to Construct Relationship Matrices for Autotetraploid and Diploid Species: A Blueberry Example. <em>The Plant Genome</em> 9(30). <a href="http://dx.doi.org/10.3835/plantgenome2016.01.0009">doi: 10.3835/plantgenome2016.01.0009</a></p>
</div>
<div id="installing-the-package" class="section level2">
<h2>Installing the Package</h2>
<p>Within R:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Installing and loading devtools
<span class="kw">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)
<span class="kw">library</span>(devtools)

## Installing from github the fullsibQTL
<span class="kw">install_github</span>(<span class="st">&quot;prmunoz/AGHmatrix&quot;</span>)</code></pre></div>
<p>To load the package</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(AGHmatrix)</code></pre></div>
</div>
<div id="relationship-matrices-using-the-pedigree-data---a-matrix" class="section level2">
<h2>Relationship matrices using the pedigree data - A matrix</h2>
<p>In this section we presented how to load the data into the software and how to construct the pedigree-based relationship matrix (A-matrix) for diploid and autotetraploid species. In the package, the function Amatrix handles the pedigree and build the A-matrix related to that given pedigree. The matrix is build according to the recursive method presented in Mrode (2014) and described by Henderson (1976). This method is expanded for higher ploidy (n-ploidy) according with Kerr et al. (2012). After loading the package you have to load your data file into the software. To do this, you can use the function <code>read.data()</code> or <code>read.csv()</code> (If specifically the format <code>.csv</code> file is used) for example. Your data should be available in R as a <code>dataframe</code> structure in the following order: column 1 must be the individual/genotype names (id), columns 2 and 3 must be the parent names. In the package there is a pedigree data example (<code>ped.mrode</code>) that can be used to look at the structure and order the data.</p>
<p>To load <code>ped.mrode</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(ped.mrode)
ped.mrode
<span class="co">#&gt;    Ind Par1 Par2</span>
<span class="co">#&gt; 1 Anc1    0    0</span>
<span class="co">#&gt; 2 Anc2    0    0</span>
<span class="co">#&gt; 3 Var1 Anc1 Anc2</span>
<span class="co">#&gt; 4 Var2 Anc1    0</span>
<span class="co">#&gt; 5 Var3 Var2 Var1</span>
<span class="co">#&gt; 6 Var4 Var3 Anc2</span></code></pre></div>
<p>The example <code>ped.mrode</code> has 3 columns, column 1 contains the names of the individual/genotypes, column 2 contains the names of the first parent, column 3 contains the names of the second parental. There is no header and the unknown value must be equal 0. Your data has to be in the same format of <code>ped.mrode</code>. In the algorithm, the first step is the pre- processing of the pedigree: the individuals are numerated <span class="math inline">\(1\)</span> to <span class="math inline">\(n\)</span>. Then, it is verified whether the genotypes in the pedigree are in chronological order (i.e. if the parents of a given individual are located prior to this individual in the pedigree dataset). If this order is not followed, the algorithm performs the necessary permutations to correct them. After this pre-processing, the matrices computation proceeds as in Mrode (2014) for diploid - for additive or dominance relationship - and as in Kerr et al. (2012) for autotetraploids - for additive relationship. For autotetraploids there is the option to include double-reduction fraction (as Slater et al., 2014). For diploids there is the option to compute the non-additive relationship matrix (Cockerham, 1954).</p>
<p>Computing additive relationship matrix for diploids:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">Amatrix</span>(ped.mrode, <span class="dt">ploidy=</span><span class="dv">2</span>)
<span class="co">#&gt; Verifying conflicting data... </span>
<span class="co">#&gt; Organizing data... </span>
<span class="co">#&gt; Your data was chronologically organized with success. </span>
<span class="co">#&gt; Constructing matrix A using ploidy = 2 </span>
<span class="co">#&gt; Completed! Time = 0  minutes</span>
<span class="co">#&gt;      Anc1  Anc2   Var1   Var2   Var3   Var4</span>
<span class="co">#&gt; Anc1 1.00 0.000 0.5000 0.5000 0.5000 0.2500</span>
<span class="co">#&gt; Anc2 0.00 1.000 0.5000 0.0000 0.2500 0.6250</span>
<span class="co">#&gt; Var1 0.50 0.500 1.0000 0.2500 0.6250 0.5625</span>
<span class="co">#&gt; Var2 0.50 0.000 0.2500 1.0000 0.6250 0.3125</span>
<span class="co">#&gt; Var3 0.50 0.250 0.6250 0.6250 1.1250 0.6875</span>
<span class="co">#&gt; Var4 0.25 0.625 0.5625 0.3125 0.6875 1.1250</span></code></pre></div>
<p>It follows some usage examples with the <code>ped.mrode</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Computing non-additive relationship matrix considering diploidy:</span>
<span class="kw">Amatrix</span>(ped.mrode, <span class="dt">ploidy=</span><span class="dv">2</span>, <span class="dt">dominance=</span><span class="ot">TRUE</span>)

<span class="co">#Computing additive relationship matrix considering autotetraploidy:</span>
<span class="kw">Amatrix</span>(ped.mrode, <span class="dt">ploidy=</span><span class="dv">4</span>)

<span class="co">#Computing additive relationship matrix considering autooctaploidy:</span>
<span class="kw">Amatrix</span>(ped.mrode, <span class="dt">ploidy=</span><span class="dv">8</span>)

<span class="co">#Computing additive relationship matrix considering autotetraploidy and double-reduction of 10%:</span>
<span class="kw">Amatrix</span>(ped.mrode, <span class="dt">ploidy=</span><span class="dv">4</span>, <span class="dt">w=</span><span class="fl">0.1</span>)

<span class="co">#Computing additive relationship matrix considering autohexaploidy and double-reduction of 10%:</span>
<span class="kw">Amatrix</span>(ped.mrode, <span class="dt">ploidy=</span><span class="dv">6</span>, <span class="dt">w=</span><span class="fl">0.1</span>)</code></pre></div>
<p>More information about <code>Amatrix</code> can be found with:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">?Amatrix</code></pre></div>
</div>
<div id="relationship-matrices-using-the-molecular-data---g-matrix-for-diploids" class="section level2">
<h2>Relationship matrices using the molecular data - G matrix for diploids</h2>
<p>This section presents how to load the data and how to construct the genomic-based relationship matrix for diploid species. In the package, the function Gmatrix is the one that handles the molecular-markers matrix and builds the relationship matrix. Molecular markers data should be organized in a matrix format (individual in rows and markers in columns) coded as 0,1,2 and missing data value (numeric or NA). Import your molecular marker data into R with the function read.table() and convert to a matrix format with the function as.matrix(). The function Gmatrix can be used then to construct the additive relationship either as proposed by Yang et al. (2010) or the proposed by VanRaden (2008). The function can also construct the dominance relationship matrix either as proposed by Su et al. (2012) or as proposed by Vitezica et al. (2013). As an example, here we build the four matrices using real data from Resende et al. (2012).</p>
<p>To load <code>snp.pine</code> and to check its structure:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(snp.pine)
<span class="kw">str</span>(snp.pine)
<span class="co">#&gt;  int [1:926, 1:4853] -9 2 2 2 2 2 2 2 2 2 ...</span>
<span class="co">#&gt;  - attr(*, &quot;dimnames&quot;)=List of 2</span>
<span class="co">#&gt;   ..$ : chr [1:926] &quot;1087120&quot; &quot;1085618&quot; &quot;1091040&quot; &quot;1091686&quot; ...</span>
<span class="co">#&gt;   ..$ : chr [1:4853] &quot;0-10024-01-114&quot; &quot;0-10037-01-257&quot; &quot;0-10040-02-394&quot; &quot;0-10040-02-41&quot; ...</span></code></pre></div>
<p>In this dataset we have 926 individuals with 4853 markers and the missing data value is <code>-9</code>.</p>
<p>It follows some usage examples with the <code>snp.pine</code> data where the unknown value (<code>missingValue</code>) is <code>-9</code>. Here we set minimum allele frequency to <code>0.05</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Computing the additive relationship matrix based on VanRaden 2008</span>
G_VanRaden &lt;-<span class="st"> </span><span class="kw">Gmatrix</span>(<span class="dt">SNPmatrix=</span>snp.pine, <span class="dt">missingValue=</span><span class="op">-</span><span class="dv">9</span>, <span class="dt">maf=</span><span class="fl">0.05</span>, <span class="dt">method=</span><span class="st">&quot;VanRaden&quot;</span>)

<span class="co">#Computing the additive relationship matrix based on Yang 2010</span>
G_Yang &lt;-<span class="st"> </span><span class="kw">Gmatrix</span>(<span class="dt">SNPmatrix=</span>snp.pine, <span class="dt">missingValue=</span><span class="op">-</span><span class="dv">9</span>, <span class="dt">maf=</span><span class="fl">0.05</span>, <span class="dt">method=</span><span class="st">&quot;Yang&quot;</span>)

<span class="co">#Computing the dominance relationship matrix based on Su 2012</span>
G_Su &lt;-<span class="st"> </span><span class="kw">Gmatrix</span>(<span class="dt">SNPmatrix=</span>snp.pine, <span class="dt">missingValue=</span><span class="op">-</span><span class="dv">9</span>, <span class="dt">maf=</span><span class="fl">0.05</span>, <span class="dt">method=</span><span class="st">&quot;Su&quot;</span>)

<span class="co">#Computing the dominance relationship matrix based on Vitezica 2013</span>
G_Vitezica &lt;-<span class="st"> </span><span class="kw">Gmatrix</span>(<span class="dt">SNPmatrix=</span>snp.pine, <span class="dt">missingValue=</span><span class="op">-</span><span class="dv">9</span>, <span class="dt">maf=</span><span class="fl">0.05</span>, <span class="dt">method=</span><span class="st">&quot;Vitezica&quot;</span>)</code></pre></div>
<p>More information about <code>Gmatrix</code> can be found with:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">?Gmatrix</code></pre></div>
</div>
<div id="relationship-matrices-using-the-molecular-data---g-matrix-for-autopolyploids" class="section level2">
<h2>Relationship matrices using the molecular data - G matrix for autopolyploids</h2>
<p>This section presents how to load the data and how to construct the genomic-based relationship matrix for autopolyploid species. In the package, the function Gmatrix is the one that handles the molecular-markers matrix and builds the relationship matrix. Molecular markers data should be organized in a matrix format (individual in rows and markers in columns) coded according to the dosage level: 0,1,2,…,ploidy level, and missing data value (numeric or NA). As an example, an autohexaploid should be coded as 0,1,2,3,4,5,6, and missing data value. Import your molecular marker data into R with the function read.table() and convert to a matrix format with the function as.matrix(). In autopolyploids, the function Gmatrix can be used then to construct: i) the additive relationship based on VanRaden (2008) and extended by Ashraf (2016) and also implemented in the <code>sommer</code> package (Covarrubias-Pazaran 2016); ii) the full-autopolyploid including additive and non-additive model as equations 8 and 9 in Slater et al. (2016); iii) the pseudo-diploid model as equations 5, 6, and 7 Slater et al. (2016). As an example, here we build the three matrices using simulated data.</p>
<p>Simulating data with 10 individuals and 100 markers.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">inds &lt;-<span class="st"> </span><span class="dv">10</span>
markers &lt;-<span class="st"> </span><span class="dv">100</span>
markersdata &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">sample</span>(<span class="dt">x=</span><span class="dv">0</span><span class="op">:</span><span class="dv">4</span>, <span class="dt">size=</span>inds<span class="op">*</span>markers, <span class="dt">replace=</span><span class="ot">TRUE</span>), <span class="dt">nrow=</span>inds, <span class="dt">ncol=</span>markers)</code></pre></div>
<p>Computing the matrices:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Computing the additive relationship matrix based on VanRaden 2008 and adapted by Ashraf 2016</span>
G_VanRaden &lt;-<span class="st"> </span><span class="kw">Gmatrix</span>(markersdata, <span class="dt">method=</span><span class="st">&quot;VanRaden&quot;</span>, <span class="dt">ploidy=</span><span class="dv">4</span>)

<span class="co">#Computing the dominance (digenic) matrix based on Endelman 2018 (Eq. 19)</span>
G_Digenic &lt;-<span class="st"> </span><span class="kw">Gmatrix</span>(markersdata, <span class="dt">method=</span><span class="st">&quot;Endelman&quot;</span>, <span class="dt">ploidy=</span><span class="dv">4</span>)

<span class="co">#Computing the full-autopolyploid matrix based on Slater 2016 (Eq. 8 and 9)</span>
G_FullAutopolyploid &lt;-<span class="st"> </span><span class="kw">Gmatrix</span>(markersdata, <span class="dt">method=</span><span class="st">&quot;Slater&quot;</span>, <span class="dt">ploidy=</span><span class="dv">4</span>)

<span class="co">#Computing the pseudodiploid matrix based on Slater 2016 (Eq. 5, 6, and 7)</span>
G_Pseudodiploid &lt;-<span class="st"> </span><span class="kw">Gmatrix</span>(markersdata, <span class="dt">method=</span><span class="st">&quot;VanRaden&quot;</span>, <span class="dt">ploidy=</span><span class="dv">4</span>, <span class="dt">pseudo.diploid=</span><span class="ot">TRUE</span>) </code></pre></div>
<p>More information about <code>Gmatrix</code> can be found with:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">?Gmatrix</code></pre></div>
</div>
<div id="combined-relationship-matrix---h-matrix" class="section level2">
<h2>Combined relationship matrix - H matrix</h2>
<p>Here we present how to compute the H matrix using the package. This matrix is computed based on the matrix A using matrix G as a correction and it is presented in Munoz et al. 2014. In this example we consider the A matrix as an identity (<em>e.g</em> we do not have pedigree information and assume the individuals in Hardy-Weinberg equilibrium). It is important to pass to <code>Hmatrix</code> the same arguments of <code>missingValue</code> and <code>maf</code> of <code>Gmatrix</code>. If the <code>Hmat</code> gives singularity problem, you can try to add a small constant to its calculus changing argument <code>c</code> to a small value (<em>e.g</em>, <code>0.001</code>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Computing the additive relationship matrix based on Yang 2010</span>
Gmat &lt;-<span class="st"> </span><span class="kw">Gmatrix</span>(<span class="dt">SNPmatrix=</span>snp.pine, <span class="dt">missingValue=</span><span class="op">-</span><span class="dv">9</span>, <span class="dt">maf=</span><span class="fl">0.05</span>, <span class="dt">method=</span><span class="st">&quot;VanRaden&quot;</span>)

<span class="co">#Setting an A matrix structure</span>
Amat &lt;-<span class="st"> </span><span class="kw">diag</span>(<span class="dv">926</span>) <span class="co">#we have 926 individuals</span>

<span class="co">#A matrix has to have the same row and col names of G</span>
<span class="kw">rownames</span>(Amat) &lt;-<span class="st"> </span><span class="kw">colnames</span>(Amat) &lt;-<span class="st"> </span><span class="kw">rownames</span>(Gmat)

<span class="co">#Computing H matrix</span>
Hmat &lt;-<span class="st"> </span><span class="kw">Hmatrix</span>(<span class="dt">A=</span>Amat, <span class="dt">G=</span>Gmat, <span class="dt">markers=</span>snp.pine, <span class="dt">missingValue=</span><span class="op">-</span><span class="dv">9</span>, <span class="dt">maf=</span><span class="fl">0.05</span>)</code></pre></div>
</div>
<div id="covariance-matrices-due-to-epistatic-terms" class="section level2">
<h2>Covariance matrices due to epistatic terms</h2>
<p>Here we present how to easily compute the epistasis relationship matrices using Hadamard products (i.e. cell-by-cell product), denoted by <code>*</code>. For more information please see Muñoz et al. (2014). In this example we are using the molecular-based relationship matrix. First, build the additive and dominance matrices:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">A &lt;-<span class="st"> </span><span class="kw">Gmatrix</span>(<span class="dt">SNPmatrix=</span>snp.pine, <span class="dt">method=</span><span class="st">&quot;VanRaden&quot;</span>, <span class="dt">missingValue=</span><span class="op">-</span><span class="dv">9</span>, <span class="dt">maf=</span><span class="fl">0.05</span>)
D &lt;-<span class="st"> </span><span class="kw">Gmatrix</span>(<span class="dt">SNPmatrix=</span>snp.pine, <span class="dt">method=</span><span class="st">&quot;Vitezica&quot;</span>, <span class="dt">missingValue=</span><span class="op">-</span><span class="dv">9</span>,<span class="dt">maf=</span><span class="fl">0.05</span>)</code></pre></div>
<p>For the first degree epistatic terms:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Additive-by-Additive Interactions</span>
A_A &lt;-<span class="st"> </span>A<span class="op">*</span>A
<span class="co">#Dominance-by-Additive Interactions</span>
D_A &lt;-<span class="st"> </span>D<span class="op">*</span>A
<span class="co">#Dominance-by-Dominance Interactions</span>
D_D &lt;-<span class="st"> </span>D<span class="op">*</span>D</code></pre></div>
<p>For the second degree epistatic terms:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Additive-by-Additive-by-Additive Interactions</span>
A_A_A &lt;-<span class="st"> </span>A<span class="op">*</span>A<span class="op">*</span>A
<span class="co">#Additive-by-Additive-by-Dominance Interactions</span>
A_A_D &lt;-<span class="st"> </span>A<span class="op">*</span>A<span class="op">*</span>D
<span class="co">#Additive-by-Dominance-by-Dominance Interactions</span>
A_D_D &lt;-<span class="st"> </span>A<span class="op">*</span>D<span class="op">*</span>D
<span class="co">#Dominance-by-Dominance-by-Dominance Interactions</span>
D_D_D &lt;-<span class="st"> </span>D<span class="op">*</span>D<span class="op">*</span>D</code></pre></div>
<p>And so on…</p>
</div>
<div id="inverting-your-matrix-to-be-used-in-mixed-models-equations" class="section level2">
<h2>Inverting your matrix to be used in mixed models equations:</h2>
<p>To invert your matrix you can use the <code>R</code> base function <code>solve</code> if your matrix has an unique inverse:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Computing the matrix</span>
A &lt;-<span class="st"> </span><span class="kw">Amatrix</span>(<span class="dt">data=</span>ped.mrode, <span class="dt">ploidy=</span><span class="dv">4</span>, <span class="dt">w=</span><span class="fl">0.1</span>)

<span class="co">#Building its unique inverse</span>
Ainv &lt;-<span class="st"> </span><span class="kw">solve</span>(A)</code></pre></div>
<p>However, it is common a genomic-based relationship matrix not be positive-definite and, therefore, has non-unique inverse. In such case, you may use one generalized inverse computation or cholesky decomposition. Here we show how to compute the Moore-Penrose inverse implemented in the <code>MASS</code> package:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Computing the matrix</span>
Gmat &lt;-<span class="st"> </span><span class="kw">Gmatrix</span>(<span class="dt">SNPmatrix=</span>snp.pine, <span class="dt">missingValue=</span><span class="op">-</span><span class="dv">9</span>, <span class="dt">maf=</span><span class="fl">0.05</span>, <span class="dt">method=</span><span class="st">&quot;VanRaden&quot;</span>)

<span class="co">#Building its Moore-Penrose generalized inverse</span>
Ginv &lt;-<span class="st"> </span>MASS<span class="op">::</span><span class="kw">ginv</span>(Gmat)</code></pre></div>
<p>If you are still having trouble to invert the polyploid A matrix we recommend you to take a look at Hamilton &amp; Kerr (2017).</p>
</div>
<div id="exporting-your-data-to-be-used-in-asreml---csv-format" class="section level2">
<h2>Exporting your data to be used in ASReml - csv format</h2>
<p>In this section, we present how to use the function formatmatrix to export a recently build matrix in the format compatible with ASReml standalone version. That is the lower diagonal matrix formatted in three columns in .csv format (other ASCII extension could be used as well). In order to do this, we need to build a matrix, its inverse, and export it using <code>formatmatrix</code> function. ASReml can invert the relationship matrix as well, probably more efficiently than R for large matrices (i.e. <code>solve()</code> function), so no need to invert the matrix in R if matrix is large. This function has as options: <code>round.by</code>, which let you decide the number of decimals you want; <code>exclude.0</code>, if <code>TRUE</code>, remove all the zeros from your data; and, name that defines the name to be used in the exported file. Use the default if not sure what parameter use in these function. Here an example using <code>ped.mrode</code> data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Loading the data example</span>
<span class="kw">data</span>(ped.mrode)

<span class="co">#Computing the matrix</span>
A &lt;-<span class="st"> </span><span class="kw">Amatrix</span>(<span class="dt">data=</span>ped.mrode, <span class="dt">ploidy=</span><span class="dv">4</span>, <span class="dt">w=</span><span class="fl">0.1</span>)

<span class="co">#Building its inverse</span>
Ainv &lt;-<span class="st"> </span><span class="kw">solve</span>(A)

<span class="co">#Exporting it. The function &quot;formatmatrix&quot; will convert it and save in your working directory</span>
<span class="kw">formatmatrix</span>(Ainv, <span class="dt">round.by=</span><span class="dv">12</span>, <span class="dt">exclude.0=</span><span class="ot">TRUE</span>, <span class="dt">name=</span><span class="st">&quot;Ainv&quot;</span>)</code></pre></div>
</div>
<div id="making-a-loop-in-order-to-get-several-matrices" class="section level2">
<h2>Making a loop in order to get several matrices</h2>
<p>In this section, we present a simple <code>for</code> function for the user to be able to obtain in a practical way several matrices for different double reduction values (if polyploidy) to later be used in <code>ASReml</code> (for example). Similar to one made in Amadeu et al. (2016).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Loading the data example</span>
<span class="kw">data</span>(ped.mrode)

<span class="co">#Determining a double reduction range</span>
double.red&lt;-<span class="kw">seq</span>(<span class="dv">0</span>,<span class="fl">0.2</span>,<span class="fl">0.05</span>)

<span class="co">#Extracting the length of double.red</span>
n&lt;-<span class="kw">length</span>(double.red)

<span class="co">#Making the loop</span>
<span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n){
  A&lt;-<span class="kw">Amatrix</span>(<span class="dt">data=</span>ped.mrode,
             <span class="dt">ploidy=</span><span class="dv">4</span>,
             <span class="dt">w=</span>double.red[i])
  <span class="co">#Computing the inverse</span>
  A.inv&lt;-<span class="kw">solve</span>(A)
  
  <span class="co">#Exporting as csv</span>
  <span class="kw">formatmatrix</span>(<span class="dt">data=</span>A.inv,
               <span class="dt">name=</span><span class="kw">paste</span>(<span class="st">&quot;Ainv_&quot;</span>,double.red[i],<span class="dt">sep=</span><span class="st">&quot;&quot;</span>),
               <span class="dt">round.by=</span><span class="dv">12</span>,
               <span class="dt">exclude.0=</span><span class="ot">TRUE</span>)
}</code></pre></div>
</div>
<div id="bibliography" class="section level2">
<h2>Bibliography</h2>
<p>Amadeu, R. R., C. Cellon, J. W. Olmstead, A. A. Garcia, M. F. Resende, and P. R. Muñoz, 2016 AGHmatrix: R package to construct relationship matrices for autotetraploid and diploid species: a blueberry example. <em>The Plant Genome</em> 9.</p>
<p>Ashraf, B. H., S. Byrne, D. Fé, A. Czaban, T. Asp, M. G. Pedersen, I. Lenk, N. Roulund, T. Didion, C. S. Jensen, et al., 2016 Estimating genomic heritabilities at the level of family-pool samples of perennial ryegrass using genotyping-by-sequencing. <em>Theoretical and Applied Genetics</em> 129: 45-52.</p>
<p>Covarrubias-Pazaran, G. 2016 Genome-assisted prediction of quantitative traits using the R package sommer. <em>PloS one</em> 11.6: e0156744.</p>
<p>Endelman, J. B., et al., 2018. Genetic variance partitioning and genome-wide prediction with allele dosage information in autotetraploid potato. <em>Genetics</em>, 209(1) pp. 77-87.</p>
<p>Hamilton, M. G., R. J. Kerr, 2017 Computation of the inverse additive relationship matrix for autopolyploid and multiple-ploidy populations. <em>Theoretical and Applied Genetics</em>. <a href="https://doi.org/10.1007/s00122-017-3041-y" class="uri">https://doi.org/10.1007/s00122-017-3041-y</a></p>
<p>Henderson, C., 1976 A simple method for computing the inverse of a numerator relationship matrix used in prediction of breeding values. <em>Biometrics</em> pp. 69–83.</p>
<p>Kerr, R. J., L. Li, B. Tier, G. W. Dutkowski, and T. A. McRae, 2012 Use of the numerator relation ship matrix in genetic analysis of autopolyploid species. <em>Theoretical and Applied Genetics</em> 124: 1271–1282.</p>
<p>Mrode, R. A., 2014 <em>Linear models for the prediction of animal breeding values</em>. Cabi. 3rd ed.</p>
<p>Muñoz, P. R., M. F. Resende, S. A. Gezan, M. D. V. Resende, G. de los Campos, M. Kirst, D. Huber,and G. F. Peter, 2014 Unraveling additive from nonadditive effects using genomic relationship matrices. <em>Genetics</em> 198: 1759–1768.</p>
<p>R Core Team, 2016 <em>R</em>: A Language and Environment for Statistical Computing. R Foundation for Statistical Computing, Vienna, Austria.</p>
<p>Resende, M. F., P. Muñoz, M. D. Resende, D. J. Garrick, R. L. Fernando, J. M. Davis, E. J. Jokela ,T. A. Martin, G. F. Peter, and M. Kirst, 2012 Accuracy of genomic selection methods in a standard data set of loblolly pine (pinus taeda l.). <em>Genetics</em> 190: 1503–1510.</p>
<p>Slater, A. T., G. M. Wilson, N. O. Cogan, J. W. Forster, and B. J. Hayes, 2014 Improving the analysis of low heritability complex traits for enhanced genetic gain in potato. <em>Theoretical and applied genetics</em> 127: 809–820.</p>
<p>Su, G., O. F. Christensen, T. Ostersen, M. Henryon, and M. S. Lund, 2012 Estimating additive and non-additive genetic variances and predicting genetic merits using genome-wide dense single nucleotide polymorphism markers. <em>PloS one</em> 7: e45293.</p>
<p>VanRaden, P., 2008 Efficient methods to compute genomic predictions. <em>Journal of dairy science</em> 91: 4414–4423.</p>
<p>Vitezica, Z. G., L. Varona, and A. Legarra, 2013 On the additive and dominant variance and covariance of individuals within the genomic selection scope. <em>Genetics</em> 195: 1223–1230.</p>
<p>Yang, J., B. Benyamin, B. P. McEvoy, S. Gordon, A. K. Henders, D. R. Nyholt, P. A. Madden, A. C. Heath, N. G. Martin, G. W. Montgomery, et al., 2010 Common snps explain a large proportion of the heritability for human height. <em>Nature genetics</em> 42: 565–569.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()
<span class="co">#&gt; R version 3.4.4 (2018-03-15)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 16.04.4 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS: /usr/lib/libblas/libblas.so.3.6.0</span>
<span class="co">#&gt; LAPACK: /usr/lib/lapack/liblapack.so.3.6.0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] rmarkdown_1.9      knitr_1.20         devtools_1.13.5   </span>
<span class="co">#&gt; [4] AGHmatrix_0.0.3006</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] Rcpp_0.12.17     codetools_0.2-15 digest_0.6.15    withr_2.1.2     </span>
<span class="co">#&gt;  [5] rprojroot_1.3-2  backports_1.1.2  magrittr_1.5     evaluate_0.10.1 </span>
<span class="co">#&gt;  [9] stringi_1.2.2    tools_3.4.4      stringr_1.3.1    yaml_2.1.19     </span>
<span class="co">#&gt; [13] compiler_3.4.4   memoise_1.1.0    htmltools_0.3.6</span></code></pre></div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
